codalab/worker/file_util.py

```diff
@@ -200,11 +205,65 @@ def do_unzip(filename):
         do_unzip(fileobj_or_name)
 
 
-def open_file(file_path, mode='r', compression_type=CompressionTypes.UNCOMPRESSED):
+def open_file(
+    file_path, mode='r', compression_type=CompressionTypes.UNCOMPRESSED, sub_zip_directory=False
+):
     """
-    Opens the given file. Can be in a directory.
+    Opens the file indicated by the given file path. Can be in a directory.
+    sub_zip_directory: If True, indicates that the given file path points to a directory that exists
+        within a zip file on Azure Blob Storage, in which case that directory will be downloaded,
+        re-zipped, and returned by this function. Default is False.
+    """
+    linked_bundle_path = parse_linked_bundle_url(file_path)
+    if (
+        linked_bundle_path.uses_beam
+        and linked_bundle_path.is_zip
+        and linked_bundle_path.zip_subpath
+    ):
+        # If file path is a zip file on Azure, open the specified path within the
+        # zip file. zipfile.open only supports 'r', not 'rb', so we always pass in 'r'.
+        with SeekableZipFile(
+            FileSystems.open(linked_bundle_path.bundle_path, compression_type)
+        ) as f:
+            zinfo = f.getinfo(linked_bundle_path.zip_subpath + ("/" if sub_zip_directory else ""))
+            if zinfo.is_dir():
+                # If streaming a folder within an Azure bundle, we need to download its contents,
+                # re-archive the folder, and return the .tar.gz version of that folder.
```

Comment:
Note that we have to extract and re-zip parts of the archive if the user is streaming a particular folder that is located within a bundle.